[RF-DOCS] Active Record Migration Documentation (#51928)
rails/rails: guides/source/active_record_migrations.md

Migrations are a feature of Active Record that allows you to evolve your
database schema over time. Rather than write schema modifications in pure SQL,
migrations allow you to use a Ruby Domain Specific Language (DSL) to describe
changes to your tables.

After reading this guide, you will know:

* Which generators you can use to create migrations.
* Which methods Active Record provides to manipulate your database.
* How to change existing migrations and update your schema.
* How migrations relate to `schema.rb`.
* How to maintain referential integrity.

--------------------------------------------------------------------------------

Migration Overview
------------------

Migrations are a convenient way to [evolve your database schema over

> time](https://en.wikipedia.org/wiki/Schema_migration) in a consistent way. They

Reviewer: "Consistency" can refer to other things in databases, like data being consistent with constraint definitions.

I think the intention here is "applying the migration consistently to each Rails environment database" or something similar. Wondering if just the word alone could be improved on with more detail.

Is there a section that could be linked to that describes how schema_migrations works and how that's used to ensure versions are applied consistently?

use a Ruby [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) so that

> you don't have to write SQL by hand, allowing your schema and changes to be

Reviewer 2: I know that `SQL` as abbreviation may be more popular than `DSL` but still, it's abbreviation so perhaps it would be useful to add wiki link to this one as well

```suggestion
you don't have to write [SQL](https://en.wikipedia.org/wiki/SQL) by hand, allowing your schema and changes to be
```

database independent. We recommend that you read the guides for [Active Record
Basics](active_record_basics.html) and the [Active Record
Associations](association_basics.html) to learn more about some of the concepts
mentioned here.

You can think of each migration as being a new 'version' of the database. A
schema starts off with nothing in it, and each migration modifies it to add or

> remove tables, columns, or entries. Active Record knows how to update your

Reviewer: I'd probably cut "entries" and maybe change this to "add or remove things like tables, columns, or indexes"

Reviewer 3: Or give it even more space to explain what the "rails way" to add or remove records using migrations is, since not everyone agrees, and people may have found conflicting information when googeling.

```ruby
# db/schema.rb
ActiveRecord::Schema[8.0].define(version: 2024_05_02_100843) do
  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"

  create_table "products", force: :cascade do |t|
    t.string "name"
    t.text "description"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end
end
```

We define the change that we want to happen moving forward in time. Before this
migration is run, there will be no table. After it is run, the table will exist.
Active Record knows how to reverse this migration as well; if we roll this
migration back, it will remove the table. You can read more about rolling back
migrations [here](#rolling-back).

After defining the change that we want to occur moving forward in time, it's
essential to consider the reversibility of the migration. While Active Record
can manage the forward progression of the migration, ensuring the creation of
the table, the concept of reversibility becomes crucial. With reversible
migrations, not only does the migration create the table when applied, but it

> also enables smooth rollback functionality. In case of reverting the migration,

Reviewer 4: ```suggestion
also enables smooth rollback functionality. In case of reverting the migration mentioned in the above example,
```

Reviewer 4: I was referring to the migration at [line number 65](https://github.com/rails/rails/pull/51928/files/a9cb47b8db7efa0eae2c1c601412cadbf6a1b274#diff-2c180fd99974672183e98de5800facb91ae57e633415a23b00cceefbab0b366aR65). I think the doc should clearly emphasise what type of migration can be reverted. Thats why i suggested to add `mentioned in the above example`.

end
```

### Creating Associations

Active Record associations are used to define relationships between different
models in your application, allowing them to interact with each other through
their relationships and making it easier to work with related data. To learn
more about associations, you can refer to the [Association Basics
guide](association_basics.html).

One common use case for associations is creating foreign key references between
tables. The generator accepts column types such as `references` to facilitate
this process. [References](#references) are a shorthand for creating columns,
indexes, foreign keys, or even polymorphic association columns.

For example,

```bash
$ bin/rails generate migration AddUserRefToProducts user:references
```

generates the following [`add_reference`][] call:

```ruby
class AddUserRefToProducts < ActiveRecord::Migration[8.0]
  def change
    add_reference :products, :user, null: false, foreign_key: true
  end
end
```

> The above migration creates a foreign key called `user_id`in the `products`

Reviewer: Missing whitespace after backtick

table, where `user_id` is a reference to the `id` column in the `users` table.
It also creates an index for the `user_id` column. The schema looks as follows:

```ruby

>   t.uuid "user_id", null: false

Reviewer 4: I think the `type` should be `bigint`

```suggestion
  t.bigint "user_id", null: false
```

Reviewer 5: @ahmad19 In code review, it's important to state your reasons for taking a stance that contradicts what's in the PR.

Reviewer 6: Yes. It should be `bigint` here, unless someone changes the default type of foreign keys. we should also show what is `t` here since there is no table reference in the code snippet. So maybe we should show the entire `create_table` code equivalent to the schema of the products table.

  t.index ["user_id"], name: "index_products_on_user_id"
```

`belongs_to` is an alias of `references`, so the above could be alternatively
written as:

```bash
$ bin/rails generate migration AddUserRefToProducts user:belongs_to
```

> generating a migration that looks like the same as above.

Reviewer: "Looks the same" (remove "like"?)

of the name:

```bash
$ bin/rails generate migration CreateJoinTableUserProduct user product
```

will produce the following migration:

```ruby
class CreateJoinTableUserProduct < ActiveRecord::Migration[8.0]
  def change
    create_join_table :users, :products do |t|
      # t.index [:user_id, :product_id]
      # t.index [:product_id, :user_id]
    end
  end
end
```

[`add_column`]:
    https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_column
[`add_index`]:
    https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_index
[`add_reference`]:
    https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_reference
[`remove_column`]:
    https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-remove_column

### Other Generators that Create Migrations

In addition to the `migration` generator, the `model`, `resource`, and
`scaffold` generators will create migrations appropriate for adding a new model.
This migration will already contain instructions for creating the relevant

> table. If you tell Rails what columns you want, then statements for adding these

Reviewer: I'd recommend avoiding phrases like "tell Rails" and instead use phrasing like "supply parameters to the generator".

Reviewer 7: Disagree here.  I agree that "tell Rails" might lead to the question of "how?", but we want the Guides to be reader-friendly, and certainly new-developer friendly. We don't need to strip the it of all colloquialism.

Reviewer 4: Or may be something like this:

```suggestion
table. If you tell Rails (pass additional details) what columns you want, then statements for adding these
```

Reviewer: I can understand the desire to use colloquial verbs like "tell" to create a more gentle style, especially for less experienced folks. 

I guess it depends on the audience, as mentioned.

Phrasing like "supply" or "provide" parameters might be more technical and less friendly for newcomers, but more conventional for experienced programmers.

Sounds like the colloquial style is preferred here.

end
```

TIP: For further help with generators, run `bin/rails generate --help`.
Alternatively, you can also run `bin/rails generate model --help` or `bin/rails
generate migration --help` for help with specific generators.

Updating Migrations
------------------

Once you have created your migration file using one of the generators from the
above [section](#generating-migration-files), you can update the generated
migration file in the `db/migrate` folder to define further changes you want to
make to your database schema.

### Creating a Table

> The [`create_table`][] method is one of the most fundamental, but most of the

Reviewer: The most fundamental ... seems like we're missing a word. And by fundamental would be it more clear to say most "common additions" (adding a table).

end
```

[`create_table`]:
    https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-create_table

#### Comments

You can pass the `:comment` option with any description for the table that will
be stored in the database itself and can be viewed with database administration

> tools, such as MySQL Workbench or PgAdmin III. It's highly recommended to

Reviewer 6: Agree. I don't think it is highly recommend to specify comments in migrations. If it was this Rails guide would be adding comments in all the examples we have here.

This changes the column `part_number` on products table to be a `:text` field.

NOTE: The `change_column` command is **irreversible**. To ensure your migration
can be safely reverted, you will need to provide your own `reversible`
migration. You can read more about reversible migrations
[here](#using-reversible).

Besides `change_column`, the [`change_column_null`][] and
[`change_column_default`][] methods are used to change a null constraint and
default values of a column.

```ruby
change_column_null :products, :name, false
change_column_default :products, :approved, from: true, to: false
```

This sets `:name` field on products to a `NOT NULL` column and the default value

> of the `:approved` field from true to false. Both of these changes will only be

Reviewer: Hm. If a column has null values in it, and we add a "not null" constraint, then the not null constraint won't apply, because existing rows would violate the constraint. We'd have to remove the rows or set the columns to non-null values for the constraint to apply.

For that reason I wouldn't include the not null addition in being relevant to only "only new rows" since existing rows are checked.

add_foreign_key :articles, :authors
```

The [`add_foreign_key`][] call adds a new constraint to the `articles` table.
The constraint guarantees that a row in the `authors` table exists where the

> `id` column matches the `articles.author_id` to ensure all reviewers listed in

Reviewer: From "to ensure" to the period, I feel it's redundant and the sentence is more effective without that portion.

```

INFO: Tables with composite primary keys require passing array values rather
than integer IDs to many methods. See also the [Active Record Composite Primary
Keys](active_record_composite_primary_keys.html) guide to learn more.

### Execute Arbitrary SQL

If the helpers provided by Active Record aren't enough, you can use the

> [`execute`][] method to execute arbitrary SQL. For example,

Reviewer: I wonder if we can do better than "arbitrary" which has a negative connotation.

I think of this more as "SQL statements or commands that don't have Active Record helper methods". Using execute() is how we run that SQL. Although not sure how to say that succinctly.

```ruby
class UpdateProductPrices < ActiveRecord::Migration[8.0]
  def up

>     execute "UPDATE products SET price = 'free' WHERE 1=1;"

Reviewer: I think the "1=1" condition is unnecessary here. This example is updating all rows.

  end

  def down
    execute "UPDATE products SET price = 'original_price' WHERE price = 'free';"
  end
end
```

In this example, we're updating the `price` column of the products table to
'free' for all records.

WARNING: Modifying data directly in migrations should be approached with
caution. Consider if this is the best approach for your use case, and be aware

> of potential drawbacks such as increase complexity and maintenance overhead,

Reviewer: I think this is meant to say "increased" (past tense). 

However, I wonder if this section could be more valuable by being more specific. For example, data modification statements in migrations aren't aware of table row counts in prod. Something fast in dev could be slow in prod.

For large tables, long running statements might be worth performing outside of Rails migrations, so that they can be performed during a low activity period, possibly as batched operations.

Reviewer 8: I'm not sure if data migrations is in-scope for this guide (could be a separate guide).
There also isn't a recommended Rails-way for data migrations, that I'm aware of.

provide a `STEP` parameter:

```bash
$ bin/rails db:rollback STEP=3
```

The last 3 migrations will be reverted.

In some cases where you modify a local migration and would like to rollback that
specific migration before migrating back up again, you can use the
`db:migrate:redo` command. As with the `db:rollback` command, you can use the
`STEP` parameter if you need to go more than one version back, for example:

```bash
$ bin/rails db:migrate:redo STEP=3
```

NOTE: You could get the same result using `db:migrate`. However, these are there
for convenience so that you do not need to explicitly specify the version to
migrate to.

#### Transactions

In databases that support transactions with statements that change the schema,
each migration is wrapped in a transaction.

INFO: A transaction ensures that if a migration fails partway through, any
changes that were successfully applied are rolled back, maintaining database
consistency. This means that either all operations within the transaction are
executed successfully, or none of them are, preventing the database from being
left in an inconsistent state if an error occurs during the transaction.

> If the database does not support transactions with statements that change the

Reviewer: Postgres calls this "Transactional DDL" and the major operations are all supported, outside of creating indexes concurrently. 

Here's a writeup: https://thenewstack.io/why-and-when-you-need-transactional-ddl-in-your-database/

It might be worth describing what we can expect for the other supported database adapters.

If that's too much scope for this section, I'd be happy to contribute a writeup externally that could possibly be included as an appendix or be linked to.

This can be a confusing area for developers as it relates to transactions, DDL, up/down migrations, and inside/outside transactions (concurrently).

Reviewer 8: Maybe that belongs in the Postgres guide?

Reviewer 8: The API docs also call these [DDL transactions](https://edgeapi.rubyonrails.org/classes/ActiveRecord/Migration.html#class-ActiveRecord::Migration-label-Transactional+Migrations). It also helps in understanding the `disable_ddl_transaction!` method mentioned later on.
```suggestion
If the database does not support DDL transactions with statements that change the
```

schema, then when a migration fails, the parts of it that have succeeded will
not be rolled back. You will have to rollback the changes manually.

NOTE: There are certain queries that can't run inside a transaction. If your
adapter supports DDL transactions you can use `disable_ddl_transaction!` method
to disable transactions for a single migration, allowing these queries to be

> executed outside the transactional boundary.

Reviewer: In a pedantic sense, this isn't strictly true. Even for concurrently in Postgres there's a transaction that initially commits the index definition in an invalid state. There are actually three total transactions!

I can contribute a writeup on this if it helps. It would be specific to Postgres though and is likely too much scope/too detailed for this section, but possibly as an appendix.

$ bin/rails db:rollback
```

This will rollback the latest migration, either by reverting the `change` method
or by running the `down` method. If you need to undo several migrations you can
provide a `STEP` parameter:

```bash
$ bin/rails db:rollback STEP=3
```

The last 3 migrations will be reverted.

In some cases where you modify a local migration and would like to rollback that
specific migration before migrating back up again, you can use the
`db:migrate:redo` command. As with the `db:rollback` command, you can use the
`STEP` parameter if you need to go more than one version back, for example:

```bash
$ bin/rails db:migrate:redo STEP=3
```

NOTE: You could get the same result using `db:migrate`. However, these are there
for convenience so that you do not need to explicitly specify the version to
migrate to.

#### Transactions

> In databases that support transactions with statements that change the schema,
> each migration is wrapped in a transaction.

Reviewer 8: "transactions with statements that change the schema" refers to DDL transactions (or transaction DDL).
I think it would help if we mention the term here first.
```suggestion
In databases that support DDL transactions — changing the schema in a single transaction —
each migration is wrapped in a transaction.
```

INFO: A transaction ensures that if a migration fails partway through, any
changes that were successfully applied are rolled back, maintaining database
consistency. This means that either all operations within the transaction are
executed successfully, or none of them are, preventing the database from being
left in an inconsistent state if an error occurs during the transaction.

If the database does not support transactions with statements that change the
schema, then when a migration fails, the parts of it that have succeeded will
not be rolled back. You will have to rollback the changes manually.

> NOTE: There are certain queries that can't run inside a transaction. If your
> adapter supports DDL transactions you can use `disable_ddl_transaction!` method
> to disable transactions for a single migration, allowing these queries to be
> executed outside the transactional boundary.

Reviewer 8: Maybe we can replace this NOTE with the [example from the API docs](https://edgeapi.rubyonrails.org/classes/ActiveRecord/Migration.html#class-ActiveRecord::Migration-label-Transactional+Migrations), and avoid the term "the transactional boundary".
````suggestion
There are queries that you can’t execute inside a transaction though, and for these situations you can turn the automatic transactions off with `disable_ddl_transaction!`:

```ruby
class ChangeEnum < ActiveRecord::Migration[8.0]
  disable_ddl_transaction!

  def up
    execute "ALTER TYPE model_size ADD VALUE 'new_value'"
  end
end
```
NOTE: Remember that you can still open your own transactions, even if you are in a Migration with self.disable_ddl_transaction!.
````

Author: Love this, thank you!

The `bin/rails db:reset` command will drop the database and set it up again.
This is functionally equivalent to `bin/rails db:drop db:setup`.

NOTE: This is not the same as running all the migrations. It will only use the
contents of the current `db/schema.rb` or `db/structure.sql` file. If a
migration can't be rolled back, `bin/rails db:reset` may not help you. To find
out more about dumping the schema see [Schema Dumping and You][] section.

[Schema Dumping and You]: #schema-dumping-and-you

### Running Specific Migrations

If you need to run a specific migration up or down, the `db:migrate:up` and
`db:migrate:down` commands will do that. Just specify the appropriate version
and the corresponding migration will have its `change`, `up` or `down` method
invoked, for example:

```bash
$ bin/rails db:migrate:up VERSION=20240428000000
```

By running this command the `change` method (or the `up` method) will be
executed for the migration with the version "20240428000000".

First, this command will check whether the migration exists and if it has
already been performed and if so, it will do nothing.

If the version specified does not exist, Rails will throw an exception.

```bash

> $ bin/rails db:migrate VERSION=4010120906120000

Author: This makes sense, I'll use `00000000000000` and update the line to match, thank you!

applied to the database. Rails uses this information to determine which
migrations need to be run when you run rails db:migrate or rails db:migrate:up
commands.

Changing Existing Migrations
----------------------------

Occasionally you will make a mistake when writing a migration. If you have
already run the migration, then you cannot just edit the migration and run the
migration again: Rails thinks it has already run the migration and so will do
nothing when you run `bin/rails db:migrate`. You must rollback the migration
(for example with `bin/rails db:rollback`), edit your migration, and then run
`bin/rails db:migrate` to run the corrected version.

In general, editing existing migrations that have been already committed to
source control is not a good idea. You will be creating extra work for yourself
and your co-workers and cause major headaches if the existing version of the
migration has already been run on production machines. Instead, you should write
a new migration that performs the changes you require.

However, editing a freshly generated migration that has not yet been committed
to source control (or, more generally, which has not been propagated beyond your

> development machine) is relatively harmless.

Reviewer: I'd even say "this is expected" instead of "relatively harmless" which could be misleading.

I personally never generate migrations that are exactly how I want. I generate them as a starting point then apply my own customizations, editing the file, then run it locally. Once it's all done I commit both files (migration file and schema dump changes) and would no longer edit from that point, instead creating a new migration as described.

guarantee referential integrity and so some people augment them with [foreign
key constraints][] in the database.

> In practice, foreign key constraints and unique indexes are generally considered

Reviewer: The way I've phrased it is database level constraints add "stronger consistency guarantees", since they're enforced on the server versus the client Rails app.

```

 [Engines]: engines.html

## Miscellaneous

> ### Using UUIDs instead of IDs for Primary Keys
> 
> By default, Rails uses auto-incrementing integers (often referred to as "IDs")

Reviewer 8: I haven't seen ID referred in this way either and couldn't find any examples online.
ID is pretty ambiguous and can be non auto-incrementing as well I think.
This is what Wikipedia uses for [Primary Key](https://en.wikipedia.org/wiki/Primary_key)
> Informally, a primary key is "which attributes identify a record," and in simple cases constitute a single attribute: a unique ID.

Maybe we can avoid that and use something like the following instead?
```suggestion
### Using UUIDs instead of IDs for Primary Keys

By default, Rails uses auto-incrementing integers IDs
```

> By default, Rails uses auto-incrementing integers (often referred to as "IDs")

Author: Thanks for confirming @p8 and thanks for bringing up this discussion @jgaskins! I'll update accordingly!

as primary keys for database records. However, there are scenarios where using
Universally Unique Identifiers (UUIDs) as primary keys can be advantageous,
especially in distributed systems or when integration with external services is
necessary. UUIDs provide a globally unique identifier without relying on a
centralized authority for generating IDs.

#### Enabling UUIDs in Rails

Before using UUIDs in your Rails application, you'll need to ensure that your
database supports storing them. Additionally, you may need to configure your
database adapter to work with UUIDs.

1. Database Configuration

    Ensure that your database supports storing UUIDs. PostgreSQL natively
    supports UUIDs with its `uuid` data type, while other databases might
    require additional configuration.

>     Enabling `uuid-ossp` Extension: If you're using PostgreSQL, enable the

Reviewer: I don't think this extension is required to generate uuid values in Postgres in recent versions (I can look this up) or store data using the uuid data type.

If you'd like a longer explanation I can provide that. My purpose in this comment is to avoid guidance to enable an extension that's not necessary.

Reviewer 9: Yes, as of [version 13](https://www.postgresql.org/docs/release/13.0/):

> Add function [gen_random_uuid()](https://www.postgresql.org/docs/13/functions-uuid.html) to generate version-4 UUIDs (Peter Eisentraut)

In order for the rest of the guide to make sense on older versions of Postgres, pgcrypto needs to be enabled for [`gen_random_uuid()`](https://www.postgresql.org/docs/12/pgcrypto.html#id-1.11.7.34.9). The older uuid-ossp extension provides the [`uuid_generate_v4()`](https://www.postgresql.org/docs/12/uuid-ossp.html) function, and the documentation for it even recommends using pgcrypto for UUIDv4.

Author: Thanks so much for pointing this out, I've amended it as follows: 

```
1. Database Configuration

    If you are using a version of PostgreSQL prior to 13, you may still need to
    enable the pgcrypto extension to access the `gen_random_uuid()` function. For
    PostgreSQL 13 and later, this is not necessary.

    For PostgreSQL versions prior to 13:

    ```ruby
    class EnablePgcryptoExtension < ActiveRecord::Migration[8.0]
      def change
        enable_extension "pgcrypto"
      end
    end
    ```
```

      end
    end
    ```

2. Rails Configuration

    In your Rails application configuration file (`config/application.rb`), add
    the following line to configure Rails to generate UUIDs as primary keys by
    default:

    ```ruby
    config.generators do |g|
      g.orm :active_record, primary_key_type: :uuid
    end
    ```

    This setting instructs Rails to use UUIDs as the default primary key type
    for ActiveRecord models.

3. Adding References with UUIDs:

    When creating associations between models using references, ensure that you
    specify the data type as :uuid to maintain consistency with the primary key
    type. For example:

    ``` ruby
    create_table :posts, id: :uuid do |t|
      t.references :author, type: :uuid, foreign_key: true
      # Other columns...
      t.timestamps
    end
    ```

    In this example, the `author_id` column in the posts table references the
    `id` column of the authors table. By explicitly setting the type to `:uuid`,
    you ensure that the foreign key column matches the data type of the primary
    key it references. Adjust the syntax accordingly for other associations and
    databases.

4. Migration Changes

>     When generating migrations for your models, you'll notice that it secifies

Reviewer 10: ```suggestion
    When generating migrations for your models, you'll notice that it specifies
```

Author: Oops, thank you for catching this!

    the id to be of type uuid:

    ```bash
      $ bin/rails g migration CreateAuthors
    ```

    ```ruby
    class CreateAuthors < ActiveRecord::Migration[8.0]
      def change
        create_table :authors, id: :uuid do |t|
          t.timestamps
        end
      end
    end
    ```

    which results in the following schema:

    ```ruby

>       create_table "authors", id: :uuid, default: -> { "gen_random_uuid()" }, force: :cascade do |t|
>       t.datetime "created_at", precision: 6, null: false
>       t.datetime "updated_at", precision: 6, null: false
>     end

Reviewer 11: ```suggestion
    create_table "authors", id: :uuid, default: -> { "gen_random_uuid()" }, force: :cascade do |t|
      t.datetime "created_at", precision: 6, null: false
      t.datetime "updated_at", precision: 6, null: false
    end
```

    ```

    In this migration, the `id` column is defined as a UUID primary key with a
    default value generated by the `gen_random_uuid()` function provided by the

>     `uuid-ossp` extension.

Reviewer 5: Slight errata, but the circumstances are confusing and it's easy to understand how you arrived at it. 😄

The `uuid_generate_v4()` function is provided by [`uuid-ossp`](https://www.postgresql.org/docs/current/uuid-ossp.html), but `gen_random_uuid()` was added as a baseline function in [Postgres v13](https://www.postgresql.org/docs/13/functions-uuid.html) specifically to avoid having to use an extension to generate a random one since the `UUID` data type is baseline and often used for its property of randomness.

Author: Thanks for the clarification!

I've amended it as follows since the UUID generation functions are no longer needed through the external extensions since Postgres version 13.

```
 In this migration, the `id` column is defined as a UUID primary key with a default value generated by the `gen_random_uuid()` function.
```

UUIDs are guaranteed to be globally unique across different systems, making them
suitable for distributed architectures. They also simplify integration with
external systems or APIs by providing a unique identifier that doesn't rely on
centralized ID generation, and unlike auto-incrementing integers, UUIDs don't
expose information about the total number of records in a table, which can be
beneficial for security purposes.

However, UUIDs are also larger than integers which can lead to increased storage
requirements, especially in large databases. They can also impact performance
due to their size. Therefore, it's essential to evaluate the trade-offs and
consider the specific requirements of your application before deciding to use

> UUIDs as primary keys.

Reviewer 2: Perhaps it would be useful to also add disclaimer about how using UUID (and especially fully random versions) may affect BTREE indexing performance due to rebalancing of the tree?

Reviewer: I agree. I discourage their use with my clients however some already have them in place, and the negative impacts to performance for v4 UUIDs is significant.

Fortunately for v7 UUIDs this will  go away eventually, but it's not yet standard.

I can offer a write-up on this if it would help.

I do recommend at a minimum suggesting UUIDs will definitely have worse performance for writes and reads compared with integer primary key and foreign keys, primarily by preventing the use of Btree indexes that would have otherwise been used had integer types been chosen instead.

Reviewer 9: > Fortunately for v7 UUIDs this will go away eventually, but it's not yet standard.

[RFC 9562](https://datatracker.ietf.org/doc/html/rfc9562) was published on [May 7](https://datatracker.ietf.org/doc/rfc9562/history/) so that's not a problem anymore 😄

Reviewer 8: Hmm, BTREE indexing performance specifics is probably out of scope for this guide.

The line about impacting "performance due to their size" seems like a repetition of the previous line (" UUIDs are also larger than integers which can lead to increased storage requirements, especially in large databases")

Maybe make it more generic and move the last comment to a NOTE?
```
However, UUIDs can also impact performance due to their size and are
harder to index. UUIDs will have worse performance for writes and reads
compared with integer primary keys and foreign keys. 

NOTE: Therefore, it's essential to evaluate the trade-offs and
consider the specific requirements of your application before deciding to use
UUIDs as primary keys.
```

Or maybe 
```
However, UUIDs are also larger than integers which can lead to increased storage
requirements, especially in large databases. Furthermore, UUID values don’t index
well due to their size and randomness.

NOTE: Therefore, it's essential to evaluate the trade-offs and
consider the specific requirements of your application before deciding to use
UUIDs as primary keys.
```

index. UUIDs will have worse performance for writes and reads compared with
integer primary keys and foreign keys.

NOTE: Therefore, it's essential to evaluate the trade-offs and consider the
specific requirements of your application before deciding to use UUIDs as
primary keys.

### Data Migrations

Data migrations involve transforming or moving data within your database. In
Rails, it is generally not advised to perform data migrations using migration
files. Here’s why:

- **Separation of Concerns**: Schema changes and data changes have different
  lifecycles and purposes. Schema changes alter the structure of your database,
  while data changes alter the content.
- **Rollback Complexity**: Data migrations can be hard to rollback safely and
  predictably.
- **Performance**: Data migrations can take a long time to run and may lock your
  tables, affecting application performance and availability.

Instead, consider using the
[`maintenance_tasks`](https://github.com/Shopify/maintenance_tasks) gem. This
gem provides a framework for creating and managing data migrations and other
maintenance tasks in a way that is safe and easy to manage without interfering
with schema migrations.

> 

Reviewer 8: This looks great!
